sprocs/sync_course_instances.sql

```diff
@@ -137,10 +138,11 @@ BEGIN
         FROM disk_course_instances AS src
         WHERE (src.errors IS NULL OR src.errors = '')
     ) LOOP
-        SELECT ci.id INTO syncing_course_instance_id
+        SELECT ci.id INTO STRICT syncing_course_instance_id
```

Comment:
Whoa, what does the `STRICT` keyword do?

Reply:
It ensures that the SELECT query returns exactly one row, which is what goes into `syncing_course_instance_id`. Without STRICT it allows the SELECT to return no rows (in which case `syncing_course_instance_id` is NULL) or multiple rows (in which case it picks the first one at random to put into `syncing_course_instance_id`). This last case was what was happening for us when we didn't have `WHERE deleted_at IS NULL`, and we were sometimes getting the "correct" row that we wanted, and sometimes we were getting one of the old rows.

When we use INTO with an INSERT or UPDATE then it's automatically STRICT. It's only SELECT INTO that needs this. I went through all other cases in the code base where we use SELECT INTO. They can all only return a single row by design, so I didn't bother adding STRICT to them.

Reply:
I also only learned about STRICT today 😄. I had thought that it was automatically applied (but it turns out it's only automatic for INSERT/UPDATE). See https://www.postgresql.org/docs/current/plpgsql-statements.html#PLPGSQL-STATEMENTS-SQL-ONEROW